File: project/app/src/main/java/com/iven/musicplayergo/adapters/AlbumsAdapter.java
package com.iven.musicplayergo.adapters;

import android.content.Context;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.android.material.card.MaterialCardView;
import com.iven.musicplayergo.R;
import com.iven.musicplayergo.Utils;
import com.iven.musicplayergo.models.Album;
import com.iven.musicplayergo.playback.PlayerAdapter;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.graphics.ColorUtils;
import androidx.recyclerview.widget.RecyclerView;

public class AlbumsAdapter extends RecyclerView.Adapter<AlbumsAdapter.SimpleViewHolder> {

    private final Context mContext;
    private final PlayerAdapter mPlayerAdapter;
    private final AlbumSelectedListener mAlbumSelectedListener;
    private final List<Album> mAlbums;
    private final int mAccent;
    private final boolean sThemeInverted;
    private Album mSelectedAlbum;
    private int mSelectedPosition;

    public AlbumsAdapter(@NonNull final Context context, @NonNull final PlayerAdapter playerAdapter, @NonNull final List<Album> albums, final boolean showPlayedArtist, final int accent, final boolean isThemeDark) {
        mContext = context;
        mPlayerAdapter = playerAdapter;
        mAlbums = albums;
        mAccent = accent;
        mAlbumSelectedListener = (AlbumSelectedListener) mContext;
        mSelectedAlbum = showPlayedArtist ? mPlayerAdapter.getCurrentSong().getSongAlbum() : mPlayerAdapter.getNavigationAlbum() != null ? mPlayerAdapter.getNavigationAlbum() : mAlbums.get(0);
        mSelectedPosition = mSelectedAlbum.getAlbumPosition();
        sThemeInverted = isThemeDark;
        mAlbumSelectedListener.onAlbumSelected(mSelectedAlbum);
    }

    @Override
    @NonNull
    public SimpleViewHolder onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {

        final View itemView = LayoutInflater.from(mContext)
                .inflate(R.layout.album_item, parent, false);

        return new SimpleViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull final SimpleViewHolder holder, final int position) {

        final Album album = mAlbums.get(holder.getAdapterPosition());
        final String albumTitle = album.getTitle();
        holder.title.setText(albumTitle);
        holder.year.setText(Album.getYearForAlbum(mContext, album.getYear()));
        holder.container.setCardBackgroundColor(ColorUtils.setAlphaComponent(mAccent, 25));
        if (!mSelectedAlbum.getTitle().equals(album.getTitle())) {
            Utils.setCardStroke(mContext, holder.container, Color.TRANSPARENT);
        } else {
            Utils.setCardStroke(mContext, holder.container, sThemeInverted? Color.WHITE : Color.BLACK);
            mSelectedPosition = holder.getAdapterPosition();
        }
    }

    @Override
    public int getItemCount() {
        return mAlbums.size();
    }

    public interface AlbumSelectedListener {
        void onAlbumSelected(@NonNull final Album album);
    }

    class SimpleViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        final TextView title, year;
        final MaterialCardView container;

        SimpleViewHolder(@NonNull final View itemView) {
            super(itemView);
            container = (MaterialCardView) itemView;
            title = itemView.findViewById(R.id.album);
            year = itemView.findViewById(R.id.year);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(@NonNull final View v) {

            //update songs list only if the album is updated
            if (getAdapterPosition() != mSelectedPosition) {
                notifyItemChanged(mSelectedPosition);
                mSelectedPosition = getAdapterPosition();
                Utils.setCardStroke(mContext, container, sThemeInverted? Color.WHITE : Color.BLACK);
                mSelectedAlbum = mAlbums.get(getAdapterPosition());
                mPlayerAdapter.setNavigationAlbum(mSelectedAlbum);
                mAlbumSelectedListener.onAlbumSelected(mSelectedAlbum);
            }
        }
    }
}

File: project/app/src/main/java/com/iven/musicplayergo/adapters/AlbumsAdapter.kt
package com.iven.musicplayergo.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.iven.musicplayergo.R
import com.iven.musicplayergo.music.Album
import com.iven.musicplayergo.ui.Utils
import kotlinx.android.synthetic.main.album_item.view.*

class AlbumsAdapter(
    private val albums: List<Album>,
    private val accent: Int,
    private val isThemeInverted: Boolean
) :
    RecyclerView.Adapter<AlbumsAdapter.AlbumsHolder>() {

    var onAlbumClick: ((String?) -> Unit)? = null

    private var mSelectedAlbum: String? = null
    private var mSelectedPosition = 0

    init {
        mSelectedAlbum = albums[0].title
    }

    fun swapSelectedAlbum(newSelectedPosition: Int) {
        mSelectedPosition = newSelectedPosition
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AlbumsHolder {

        val albumView =
            LayoutInflater.from(parent.context).inflate(R.layout.album_item, parent, false)
        val materialCardView = albumView as CardView
        val backgroundColor =
            if (isThemeInverted) Utils.darkenColor(accent, 0.85F) else Utils.lightenColor(
                accent,
                0.80F
            )
        materialCardView.setCardBackgroundColor(backgroundColor)

        return AlbumsHolder(albumView)
    }

    override fun getItemCount(): Int {
        return albums.size
    }

    override fun onBindViewHolder(holder: AlbumsHolder, position: Int) {
        val title = albums[holder.adapterPosition].title
        val year = albums[holder.adapterPosition].year
        holder.bindItems(title, year)
    }

    inner class AlbumsHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bindItems(title: String?, year: String?) {

            val albumCard = itemView as CardView

            /*         if (mSelectedPosition != adapterPosition) albumCard.strokeColor =
                         Color.TRANSPARENT else albumCard.strokeColor = accent*/

            itemView.album.text = title
            itemView.year.text = year
            itemView.setOnClickListener {
                if (adapterPosition != mSelectedPosition) {
                    notifyItemChanged(mSelectedPosition)
                    mSelectedPosition = adapterPosition
                    //   albumCard.strokeColor = accent
                    mSelectedAlbum = title
                    onAlbumClick?.invoke(title)
                }
            }
        }
    }
}
